### PR TITLE
[fix](storage) low_cardinality_optimize core dump when is null predicate

### LINKED ISSUE #9555: [Bug]low_cardinality_optimize core dump in dev-1.0.1
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Version

dev-1.0.1

### What's Wrong?

using low_cardinality_optimize run regression case in dev-1.0.1，BE will coredump.

**regression case**
regression-test/suites/data_model/duplicate/storage/test_dup_tab_char_nullable.groovy

**core stack**
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:49
#1  0x00007f63c4a07864 in __GI_abort () at abort.c:79
#2  0x0000562a85496fd2 in __gnu_cxx::__verbose_terminate_handler() [clone .cold] ()
#3  0x0000562a910f6466 in __cxxabiv1::__terminate(void (*)()) ()
#4  0x0000562a910f64d1 in std::terminate() ()
#5  0x0000562a910f6625 in __cxa_throw ()
#6  0x0000562a88d729cc in doris::vectorized::throwFromErrno (s=..., code=56, e=12) at /home/zcp/selectdb/be/src/vec/common/exception.cpp:115
#7  0x0000562a86002573 in Allocator<false, false>::realloc (this=0x562b97a5cda0, buf=0xffffffffffffffff, old_size=8192, new_size=9223372036854775808, alignment=0)
    at /home/zcp/selectdb/be/src/vec/common/allocator.h:145
#8  0x0000562a85fedf87 in doris::vectorized::PODArrayBase<1ul, 4096ul, Allocator<false, false>, 15ul, 16ul>::realloc<>(unsigned long) (this=0x562b97a5cda0, bytes=9223372036854775808)
    at /home/zcp/selectdb/be/src/vec/common/pod_array.h:147
#9  0x0000562a85fd8a1d in doris::vectorized::PODArrayBase<1ul, 4096ul, Allocator<false, false>, 15ul, 16ul>::reserve<>(unsigned long) (this=0x562b97a5cda0, n=7957695015409771385)
    at /home/zcp/selectdb/be/src/vec/common/pod_array.h:213
#10 0x0000562a85fc9123 in doris::vectorized::PODArrayBase<1ul, 4096ul, Allocator<false, false>, 15ul, 16ul>::resize<>(unsigned long) (this=0x562b97a5cda0, n=7957695015409771385)
    at /home/zcp/selectdb/be/src/vec/common/pod_array.h:219
#11 0x0000562a85fa4ed0 in doris::vectorized::ColumnString::insert_data (this=0x562b97a5cd70, pos=0x653a3a74736f6f62 <error: Cannot access memory at address 0x653a3a74736f6f62>,
    length=7957695015409771384) at /home/zcp/selectdb/be/src/vec/columns/column_string.h:154
#12 0x0000562a85fc0697 in doris::vectorized::ColumnDictionary<int>::filter_by_selector (this=0x562b4992d860, sel=0x7f6364cb2d50, sel_size=1, col_ptr=0x562b97a5cd70)
    at /home/zcp/selectdb/be/src/vec/columns/column_dictionary.h:205
#13 0x0000562a88c0e0c9 in doris::vectorized::ColumnNullable::filter_by_selector (this=0x562ad7a60c80, sel=0x7f6364cb2d50, sel_size=1, col_ptr=0x562bc9d898e0)
    at /home/zcp/selectdb/be/src/vec/columns/column_nullable.cpp:195
#14 0x0000562a8cb84157 in doris::vectorized::Block::copy_column_data_to_block (this=0x562b97a5c190, input_col_ptr=0x562ad7a60c80, sel_rowid_idx=0x7f6364cb2d50, select_size=1, block_cid=0,
    batch_size=4096) at /home/zcp/selectdb/be/src/vec/core/block.h:128
#15 0x0000562a8cb9239e in doris::segment_v2::SegmentIterator::_output_column_by_sel_idx<std::set<unsigned int, std::less<unsigned int>, std::allocator<unsigned int> > > (this=0x562aed4fa500,
    block=0x562b97a5c190, column_ids=..., sel_rowid_idx=0x7f6364cb2d50, select_size=1) at /home/zcp/selectdb/be/src/olap/rowset/segment_v2/segment_iterator.h:115
#16 0x0000562a8cb7d811 in doris::segment_v2::SegmentIterator::next_batch (this=0x562aed4fa500, block=0x562b97a5c190) at /home/zcp/selectdb/be/src/olap/rowset/segment_v2/segment_iterator.cpp:983
#17 0x0000562a867f4265 in doris::BetaRowsetReader::next_block (this=0x562bdabd7380, block=0x562b97a5c190) at /home/zcp/selectdb/be/src/olap/rowset/beta_rowset_reader.cpp:193
#18 0x0000562a8e3796af in doris::vectorized::VCollectIterator::Level0Iterator::next (this=0x562b26d9fc20, block=0x562b97a5c190) at /home/zcp/selectdb/be/src/vec/olap/vcollect_iterator.cpp:215
#19 0x0000562a8e383887 in doris::vectorized::VCollectIterator::Level1Iterator::_normal_next (this=0x562b81f937a0, block=0x562b97a5c190)
    at /home/zcp/selectdb/be/src/vec/olap/vcollect_iterator.cpp:366
#20 0x0000562a8e383f09 in doris::vectorized::VCollectIterator::Level1Iterator::_normal_next (this=0x562b81f937a0, block=0x562b97a5c190)
    at /home/zcp/selectdb/be/src/vec/olap/vcollect_iterator.cpp:375
#21 0x0000562a8e37ac95 in doris::vectorized::VCollectIterator::Level1Iterator::next (this=0x562b81f937a0, block=0x562b97a5c190) at /home/zcp/selectdb/be/src/vec/olap/vcollect_iterator.cpp:264
#22 0x0000562a8e3771e3 in doris::vectorized::VCollectIterator::next (this=0x562bec964408, block=0x562b97a5c190) at /home/zcp/selectdb/be/src/vec/olap/vcollect_iterator.cpp:165
#23 0x0000562a8e3945eb in doris::vectorized::BlockReader::_direct_next_block (this=0x562bec964000, block=0x562b97a5c190, mem_pool=0x0, agg_pool=0x0, eof=0x7f6364cb3648)
    at /home/zcp/selectdb/be/src/vec/olap/block_reader.cpp:175
#24 0x0000562a8e3a0af5 in doris::vectorized::BlockReader::next_block_with_aggregation (this=0x562bec964000, block=0x562b97a5c190, mem_pool=0x0, agg_pool=0x0, eof=0x7f6364cb3648)
    at /home/zcp/selectdb/be/src/vec/olap/block_reader.h:49
#25 0x0000562a8d1ebf86 in doris::vectorized::VOlapScanner::get_block (this=0x562bb7a62dc0, state=0x562c9dc99000, block=0x562b97a5c190, eof=0x7f6364cb3648)
    at /home/zcp/selectdb/be/src/vec/exec/volap_scanner.cpp:55

### What You Expected?

dev-1.0.1 will not coredump when using vectorization storage layer run regression case.

### How to Reproduce?

_No response_

### Anything Else?

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by zenoyang**
# Proposed changes

Issue Number: close #9555 
Make the last value of the dictionary null, when ColumnDict inserts a null value, add the encoding corresponding to the last value of the dictionary·

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/columns/column_dictionary.h [hunk 1]
```diff
@@ -98,12 +98,14 @@ class ColumnDictionary final : public COWHelper<IColumn, ColumnDictionary<T>> {
     }
 
     void insert_data(const char* pos, size_t /*length*/) override {
-        _codes.push_back(unaligned_load<T>(pos));
+        if (pos == nullptr) {
```
Comment (BiteTheDDDDt) [inline be/src/vec/columns/column_dictionary.h:101]: Why we should process this logic? It seems just like `insert_default`

File: be/src/vec/columns/column_dictionary.h [hunk 2]
```diff
@@ -98,12 +98,14 @@ class ColumnDictionary final : public COWHelper<IColumn, ColumnDictionary<T>> {
...
+            _codes.push_back(_dict.get_null_code());
+            return;
+        }
+        LOG(FATAL) << "insert_data not supported in ColumnDictionary";
```
Comment (dataroaring) [inline be/src/vec/columns/column_dictionary.h:105]: Should we place a DCHECK(false) here？

File: be/src/vec/columns/column_dictionary.h [hunk 3]
```diff
@@ -304,10 +310,14 @@ class ColumnDictionary final : public COWHelper<IColumn, ColumnDictionary<T>> {
             return -1;
         }
 
+        T get_null_code() {
+            return _dict_data.size() - 1; // The last dict value is null value
+        }
```
Comment (dataroaring) [inline be/src/vec/columns/column_dictionary.h:315]: For range predicate, it seems related to null order, I am not sure how doris handle null order. e.g. where col >= "xxx" includes null.

**Review by dataroaring — DISMISSED** [review]
LGTM

File: be/src/vec/columns/column_dictionary.h [hunk 4]
```diff
@@ -98,12 +98,15 @@ class ColumnDictionary final : public COWHelper<IColumn, ColumnDictionary<T>> {
     }
 
     void insert_data(const char* pos, size_t /*length*/) override {
-        _codes.push_back(unaligned_load<T>(pos));
+        DCHECK(pos == nullptr);
```
Comment (HappenLee) [inline be/src/vec/columns/column_dictionary.h:101]: if the function only dispose null value，it should not name `insert_data`

**Review by HappenLee — DISMISSED** [review]
LGTM

**Review by dataroaring — DISMISSED** [review]
LGTM

**Comment by zenoyang** [comment]
There was a problem with `get_null_code` use `dict.size()` before: when the first element read was a null value (the dictionary was not initialized).
So I changed `get_null_code` to `-1` again, please review again.
@HappenLee @dataroaring

**Comment by zenoyang** [comment]
There was a problem with `get_null_code` use `dict.size()` before: when the first element read was a null value (the dictionary was not initialized).
So I changed `get_null_code` to `-1` again, please review again.
@HappenLee @dataroaring

**Comment by dataroaring** [comment]
> 

Thanks for your great work.

Could you explain more in detail? Where the read happen in the code?

**Comment by zenoyang** [comment]
> > 
> 
> Thanks for your great work.
> 
> Could you explain more in detail? Where the read happen in the code?

The read code is in the `FileColumnIterator::next_batch(size_t* n, vectorized::MutableColumnPtr& dst, bool* has_null)` method. When `null` is read from the page, `dst->insert_data(nullptr, 0);` is called (the `insert_default` method of ColumnDict will be called).
If non-null, call `_page.data_decoder->next_batch();` (will call the `insert_many_dict_data` method of ColumnDict, the dictionary will be built on the first call)

**Review by dataroaring — APPROVED** [review]

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**